### PR TITLE
Support timeouts in the parallel runner

### PR DESCRIPTION
**Pull request by weswigham**
This allows the parallel runner to timeout tests (and thereby fail) if they exceed the specified timeout (the default 20s for now). @sandersn mentioned wanting this earlier today. On a somewhat related note, our RWC test timeout is probably waaaay too long for how long we expect RWC to take nowadays.

This implements handling of both `timeout=n` in jake, `--timeout=n` in gulp, and `this.timeout(n)` in a unittest.

### CODE REVIEW

File: Gulpfile.ts [hunk 1]
```diff
@@ -871,8 +871,17 @@ function cleanTestDirs(done: (e?: any) => void) {
...
+function writeTestConfigFile(tests: string, runners: string, light: boolean, taskConfigsFolder?: string, workerCount?: number, stackTraceLimit?: string, testTimeout?: number) {
+    const testConfigContents = JSON.stringify({
+        test: tests ? [tests] : undefined,
+        runner: runners ? runners.split(",") : undefined,
+        light,
+        workerCount,
+        stackTraceLimit,
+        taskConfigsFolder,
+        noColor: !cmdLineOptions.colors,
+        timeout: testTimeout,
```
Comment (andy-ms) [inline Gulpfile.ts]: Nit: I would just name these the same.

File: src/harness/parallel/host.ts [hunk 2]
```diff
@@ -134,13 +136,26 @@ namespace Harness.Parallel.Host {
...
             // TODO: Just send the config over the IPC channel or in the command line arguments
             const config: TestConfig = { light: Harness.lightMode, listenForWork: true, runUnitTests };
             const configPath = ts.combinePaths(taskConfigsFolder, `task-config${i}.json`);
             Harness.IO.writeFile(configPath, JSON.stringify(config));
             const child = fork(__filename, [`--config="${configPath}"`]);
+            let currentTimeout = defaultTimeout;
+            const killChild = () => {
+                child.kill();
+                console.error(`Worker exceeded timeout ${child.currentTasks && child.currentTasks.length ? `while running test '${child.currentTasks[0].file}'.` : `during test setup.`}`);
+                return process.exit(2);
```
Comment (andy-ms) [inline src/harness/parallel/host.ts]: Why return?

File: src/harness/parallel/worker.ts [hunk 3]
```diff
@@ -130,6 +152,11 @@ namespace Harness.Parallel.Worker {
                 return;
             }
             finally {
+                if (timeout !== undefined) {
```
Comment (andy-ms) [inline src/harness/parallel/worker.ts]: This looks familar. Re-use code?

File: src/harness/parallel/host.ts [hunk 4]
```diff
@@ -160,8 +175,23 @@ namespace Harness.Parallel.Host {
         Stack: ${data.payload.stack}`);
                         return process.exit(2);
                     }
+                    case "timeout": {
+                        if (data.payload.duration === -1) {
```
Comment (andy-ms) [inline src/harness/parallel/host.ts]: Use a union `number | "default"`

File: src/harness/parallel/host.ts (hunk 2 above)
Comment (weswigham) [reply src/harness/parallel/host.ts]: Just to punctuate that no code should execute while the process is exiting.

**Pull request merged**